src/solver/td3.ml

```diff
@@ -496,16 +489,19 @@ module Base =
         (* solve x Widen *)
       in
 
-      let rec destabilize_normal x =
+      let rec destabilize_normal ?(notify_destabilized = fun _ _ _ -> ()) x =
         if tracing then trace "sol2" "destabilize %a" S.Var.pretty_trace x;
         let w = HM.find_default infl x VS.empty in
         HM.replace infl x VS.empty;
         VS.iter (fun y ->
             if tracing then trace "sol2" "stable remove %a" S.Var.pretty_trace y;
+            let was_stable = HM.mem stable y in
+            let is_called = HM.mem called y in
             HM.remove stable y;
             HM.remove superstable y;
             Hooks.stable_remove y;
-            if not (HM.mem called y) then destabilize_normal y
+            notify_destabilized y is_called was_stable;
```

Comment:
It seems quite wasteful to do additional hashtable lookups whose values are completely ignored in all our usual cases when `side_widen` is not `cycle`.
`Hashtbl.mem` might seem like a cheap operation, but pre-OCaml 5.1 (which we're still on) it's unnecessarily allocating: https://github.com/ocaml/ocaml/pull/11500. This was a problem for #609 at one point.
Moreover, `destabilize` is probably the hottest code in the solver, so these changes must be done very carefully.

Reply:
The problem ist that `stable` is modified right there -- so one can not make this operation lazy easily. The question is whether passing an additional argument whether this computation is required and then checking for that before calling `mem` is really cheaper...

Reply:
Conceptually it would be possible to implement a version of `Hashtbl.remove` which returns a boolean (like hashtables do in some other languages) and does so in a single hashtable bucket lookup. But it would be horrible `BatHashtbl`-like `Obj.magic` business.

Reply:
Yes, I went looking hoping BatHashtbl might already have this, but it doesn't 😞.

Reply:
Maybe the `notify_destabilized` callback could also be called before removing y from those sets. Then, the lookup could be done in the callback as needed.